types/ot/index.d.ts

```diff
@@ -3,50 +3,477 @@
+     */
+    isRedoing(): boolean;
+}
+
+/**
+ * A WrappedOperation contains an operation and corresponing metadata.
+ */
+export class WrappedOperation<T = any> {
+    wrapped: TextOperation;
+    meta: T;
+    constructor(operation: TextOperation, meta?: T);
+    apply(doc: string): string;
+    invert(doc: string): WrappedOperation<T>;
+    compose(operation: WrappedOperation<T>): WrappedOperation<T>;
+    static transform<T>(left: WrappedOperation<T>, right: WrappedOperation<T>): WrappedOperation<T>;
+}
+
+export interface ClientAdapter {
+    registerUndo(fun: () => void): void;
+    registerRedo(fun: () => void): void;
+    getValue(): string;
+    applyOperation(operation: TextOperation): void;
+    setSelection(selection?: Selection): void;
+    getSelection(): Selection;
+    registerCallbacks(callbacks: ClientAdapterCallbacks): void;
+    setOtherSelection(selection: Selection, color: string, otherClientId: string): Mark;
+}
+
+export interface ClientAdapterCallbacks {
+    change(operation: TextOperation, inverse: TextOperation): void;
+    selectionChange(): void;
+    blur(): void;
+}
+
+export interface ServerAdapter {
+    sendSelection(selection?: Selection): void;
+    sendOperation(revision: number, operation: SerializedTextOperation, selection?: Selection): void;
+    registerCallbacks(callbacks: ServerAdapterCallbacks): void;
+}
+
+export interface ServerAdapterCallbacks {
+    client_left(clientId: string): void;
+    set_name(clientId: string, name: string): void;
+    ack(): void;
+    operation(operation: SerializedTextOperation): void;
+    selection(clientId: string, selection: string): void;
+    clients(clients: any): void;
+    reconnect(): void;
+}
+
+export interface ClientObj {
+    clientId: string;
+    name?: string;
+    selection: string;
+}
+
+export interface Clients<T = any> {
+    [clientId: string]: T;
+}
+
+export interface Mark {
+    clear(): void;
+}
+
+export class EditorClient extends Client {
+    revision: number;
+    clients: { [key: string]: EditorClient.OtherClient };
+    serverAdapter: any;
+    editorAdapter: any;
+
+    constructor(revision: number, clients: ClientObj[], serverAdapter: ServerAdapter, editorAdapter: ClientAdapter);
+
+    // not sure about all those signatures
+    addClient(clientId: string, clientObj: ClientObj): void;
+    initializeClients(clients: Clients): void;
+    getClientObject(clientId: string): ClientObj;
+    onClientLeft(clientId: string): void;
+    initializeClientList(): void;
+    applyUnredo(operation: TextOperation): void;
+    undo(): void;
+    redo(): void;
+    onChange(textOperation: TextOperation, inverse: TextOperation): void;
+    updateSelection(): void;
+    onSelectionChange(): void;
+    onBlur(): void;
+    sendSelection(selection: Selection): void;
+    sendOperation(revision: number, operation: TextOperation): void;
+    applyOperation(operation: TextOperation): void;
+}
+
+// TODO
```

Comment:
What needs to be done here?

Reply:
To describe what's inside `SelfMeta` and `OtherClient`. Maybe for the next step ?

Reply:
Personally, I don't like to leave bare TODOs in code, so I'd probably add an approximate description of the work that needs to be done.  However, I mostly wanted to confirm that you hadn't left it there unintentionally.

I agree that these don't need to be fleshed out in this change.